contracts/VeJoeStaking.sol

```diff
@@ -0,0 +1,267 @@
+contract VeJoeStaking is Initializable, OwnableUpgradeable {
+    using SafeERC20Upgradeable for IERC20Upgradeable;
+
+    struct UserInfo {
+        uint256 balance; // Amount of JOE currently staked by user
+        uint256 lastRewardTimestamp; // Timestamp of last veJOE claim, or time of first deposit if user
+        // has not claimed any veJOE yet
+        uint256 boostEndTimestamp; // Timestamp of when user stops receiving boost benefits
+    }
+
+    IERC20Upgradeable public joe;
+    VeJoeToken public veJoe;
+
+    /// @notice The maximum ratio of veJOE to staked JOE
+    /// For example, if user has `n` JOE staked, they can own a maximum of `n * maxCap` veJOE.
+    uint256 public maxCap;
+
+    /// @notice Rate of veJOE generated per sec per JOE staked
+    uint256 public baseGenerationRate;
+
+    /// @notice Boosted rate of veJOE generated per sec per JOE staked
+    uint256 public boostedGenerationRate;
+
+    /// @notice Percentage of total staked JOE user has to deposit in order to start
+    /// receiving boosted benefits, in parts per 100.
+    /// @dev Specifically, user has to deposit at least `boostedThreshold/100 * totalStakedJoe` JOE.
+    /// The only exception is the user will also receive boosted benefits if its their first
+    /// time staking.
+    uint256 public boostedThreshold;
+
+    /// @notice The length of time a user receives boosted benefits
+    uint256 public boostedDuration;
+
+    mapping(address => UserInfo) public userInfos;
+
+    event Deposit(address indexed user, uint256 amount);
+    event Withdraw(address indexed user, uint256 amount);
+    event Claim(address indexed user, uint256 amount);
+
+    /// @notice Initialize with needed parameters
+    /// @param _joe Address of the JOE token contract
+    /// @param _veJoe Address of the veJOE token contract
+    /// @param _baseGenerationRate Rate of veJOE generated per sec per JOE staked
+    /// @param _boostedGenerationRate Boosted rate of veJOE generated per sec per JOE staked
+    /// @param _boostedThreshold Percentage of total staked JOE user has to deposit to be boosted
+    /// @param _boostedDuration Length of time a user receives boosted benefits
+    function initialize(
+        IERC20Upgradeable _joe,
+        VeJoeToken _veJoe,
+        uint256 _baseGenerationRate,
+        uint256 _boostedGenerationRate,
+        uint256 _boostedThreshold,
+        uint256 _boostedDuration
+    ) public initializer {
+        require(address(_joe) != address(0), "VeJoeStaking: unexpected zero address for _joe");
+        require(address(_veJoe) != address(0), "VeJoeStaking: unexpected zero address for _veJoe");
+        require(
+            _boostedGenerationRate > _baseGenerationRate,
+            "VeJoeStaking: expected _boostedGenerationRate to be greater than _baseGenerationRate"
+        );
+        require(_boostedThreshold <= 100, "VeJoeStaking: expected _boostedThreshold to be less than or equal to 100");
```

Comment:
`_boostedThreshold` should also be greater than 0 as this would lead to division by 0 error on line 151.

Reply:
I updated the logic so that `_boostedThreshold` being 0 won't cause division by 0 error. It's also probably worth considering whether we should require it be greater than 0 anyways though